src/FSharp.Core/printf.fs

```diff
@@ -658,8 +658,28 @@ module internal PrintfImpl =
     
     /// Contains functions to handle left/right and no justification case for numbers
     module GenericNumber =
-
-        let isPositive (n: obj) = 
+        
+        let inline doubleIsPositive (n: double) =
+            n >= 0.0
+                // Ensure -0.0 is treated as negative (see https://github.com/dotnet/fsharp/issues/15557)
+                // and use bitwise comparison because floating point comparison treats +0.0 as equal to -0.0
+                && (BitConverter.DoubleToInt64Bits n <> BitConverter.DoubleToInt64Bits -0.0)
+        
+        let inline singleIsPositive (n: single) =
+            doubleIsPositive (float n)
+        
+        let decimalSignBit (n: decimal) =
+            // Unfortunately it's impossible to avoid this array allocation without either targeting .NET 5+ or relying on knowledge about decimal's internal representation
```

Comment:
`decimal's` representation is notably fixed and while it technically isn't strictly documented, it isn't really "internal".

`decimal` is a C# primitive type that must strictly be 16-bytes in size (a hard assumption exists here), but also is an interchange type for the Win32 `DECIMAL` type and it's layout must be ABI compatible with this: https://learn.microsoft.com/en-us/windows/win32/api/wtypes/ns-wtypes-decimal-r1

Reply:
We could likely explicitly document this if really needed, but it's generally safe to depend on.

There is a `decimal.IsPositive` API and you could notably also use `NativePtr.stackalloc` to avoid the allocation if the library was multitargeted

Reply:
The library is netstandard2.0/2.1, which I think rules out the IsPositive/IsNegative methods. And ._flags field itself is private, I am not sure there is a safe way to read it?

Reply:
@T-Gro we could write an equivalent struct type and use magic like [Unsafe.As](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.as?view=net-9.0#system-runtime-compilerservices-unsafe-as-2(-0@)) to bitwise cast it, but that feels evil to me

Reply:
FWIW, I would vote against doing it that way